api/graphite.go

```diff
@@ -229,9 +351,95 @@ func (s *Server) metricsFind(ctx *middleware.Context, request models.GraphiteFin
 	response.Write(ctx, response.NewJson(200, b, request.Jsonp))
 }
 
+func (s *Server) listLocal(orgId int) []schema.MetricDefinition {
+	return s.MetricIndex.List(orgId)
+}
+
+func (s *Server) listRemote(orgId int, peer *cluster.Node) ([]schema.MetricDefinition, error) {
+	log.Debug("HTTP IndexJson() querying %s/cluster/index/list for %d", peer.GetName(), orgId)
+	buf, err := peer.Post("/cluster/index/list", models.IndexList{OrgId: orgId})
+	if err != nil {
+		log.Error(4, "HTTP IndexJson() error querying %s/cluster/index/list: %q", peer.GetName(), err)
+		return nil, err
+	}
+	result := make([]schema.MetricDefinition, 0)
+	for len(buf) != 0 {
+		var def schema.MetricDefinition
+		buf, err = def.UnmarshalMsg(buf)
+		if err != nil {
+			log.Error(3, "HTTP IndexJson() error unmarshaling body from %s/cluster/index/list: %q", peer.GetName(), err)
+			return nil, err
+		}
+		result = append(result, def)
+	}
+	return result, nil
+}
+
 func (s *Server) metricsIndex(ctx *middleware.Context) {
-	list := s.MetricIndex.List(ctx.OrgId)
-	response.Write(ctx, response.NewFastJson(200, models.MetricNames(list)))
+	peers := cluster.PeersForQuery()
+	errors := make([]error, 0)
+	series := make([]schema.MetricDefinition, 0)
+	seenDefs := make(map[string]struct{})
+	var mu sync.Mutex
+	var wg sync.WaitGroup
+	for _, peer := range peers {
+		wg.Add(1)
+		if peer == cluster.ThisNode {
+			go func() {
+				result := s.listLocal(ctx.OrgId)
+				mu.Lock()
+				for _, def := range result {
+					if _, ok := seenDefs[def.Id]; !ok {
+						series = append(series, def)
+						seenDefs[def.Id] = struct{}{}
+					}
+				}
+				mu.Unlock()
+				wg.Done()
+			}()
+		} else {
+			go func() {
+				result, err := s.listRemote(ctx.OrgId, peer)
+				mu.Lock()
+				if err != nil {
+					errors = append(errors, err)
+				}
+				for _, def := range result {
+					if _, ok := seenDefs[def.Id]; !ok {
+						series = append(series, def)
+						seenDefs[def.Id] = struct{}{}
+					}
+				}
+				mu.Unlock()
+				wg.Done()
+			}()
+		}
+	}
+	wg.Wait()
+	var err error
+	if len(errors) > 0 {
+		if len(errors) == 1 {
+			err = errors[0]
+		} else {
+			msg := fmt.Sprintf("%d errors: ", len(errors))
+			for _, e := range errors {
+				msg += e.Error() + ", "
+			}
```

Comment:
instead of this loop you can just print `errors` as "%q" that will show it as a list. and in case an error message contains a `,` it will be less confusing. see https://play.golang.org/p/bjBWpMQ_R5